### PR TITLE
Apply Regex starting loop optimization to non-atomic loops as well

### PR DESCRIPTION
**Pull request by stephentoub**
@danmosemsft, @eerhardt, @pgovind, this appears sound, but it'd be great if you could double-check me on it.

### CODE REVIEW

**Comment by pgovind** [comment]
I missed this PR somehow. I'll go over the logic and comment/approve tomorrow

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs [hunk 1]
```diff
@@ -310,18 +310,18 @@ internal RegexNode FinalOptimize()
                 // to implementations that don't support backtracking.
                 EliminateEndingBacktracking(rootNode.Child(0), DefaultMaxRecursionDepth);
 
-                // Optimization: unnecessary re-processing of atomic starting groups.
-                // If an expression is guaranteed to begin with a single-character infinite atomic group that isn't part of an alternation (in which case it
+                // Optimization: unnecessary re-processing of starting loops.
+                // If an expression is guaranteed to begin with a single-character unbounded loop that isn't part of an alternation (in which case it
```
Comment (eerhardt) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:314]: `If an expression is guaranteed to begin with a single-character`

I'm missing what part checks for a single character. Can you point it out to me?

**Review by eerhardt — APPROVED** [review]
I can't think of a reason why this would break. If you are convinced it is correct, I think that's good enough for me. 😉

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:314]: I wasn't sure how to phrase it better.  By "single character", I meant loops around a single character match, e.g. One/NotOne/Set... as opposed to loops around arbitrary expressions, e.g. Loop.  So we will optimize `\w*` but not `(\w\d)*`.

**Review by pgovind — APPROVED** [review]
Alright, this looks good to me. I spent some time going over the parsing logic , ran some patterns locally and am reasonably convinced that this is a safe change.

**Pull request merged**